protocol/contracts/voting/MasterCandidate.sol

```diff
@@ -0,0 +1,191 @@
+    }
+
+    function nominate(
+        address account,
+        uint256 discordId,
+        string memory discordHandle,
+        string memory twitterHandle,
+        bytes32[] calldata merkleProof
+    ) external {
+        require(account != address(0), "MasterCandidate: Address cannot be 0x0");
+        require(MerkleProof.verify(merkleProof, merkleRoot, keccak256(abi.encode(account, discordId.toString()))), "MasterCandidate: candidate not in temple discord");
+
+        SafeERC20.safeTransferFrom(nominationFeeToken, account, address(this), nominationFee);
+        _safeMint(account, discordId);
+
+        Candidate storage candidate = candidates[discordId];
+        candidate.discordId = discordId;
+        candidate.discordHandle = discordHandle;
+        candidate.twitterHandle = twitterHandle;
+
+        emit UpdateMetadata(msg.sender, discordId, discordHandle, twitterHandle);
+    }
+
+    function updateMetadata(
+        uint256 discordId,
+        string memory discordHandle,
+        string memory twitterHandle
+    ) external {
+        require(ownerOf(discordId) == msg.sender, "Only the candidate can update metadata");
+        
+        Candidate storage candidate = candidates[discordId];
+        candidate.discordHandle = discordHandle;
+        candidate.twitterHandle = twitterHandle;
+
+        emit UpdateMetadata(msg.sender, discordId, discordHandle, twitterHandle);
+    }
+
+    function toggleEndorsement(uint256 discordId) external {
+        _toggleEndorsement(msg.sender, discordId);
+    }
+
+    function toggleEndorsements(uint256[] memory discordIds) external {
+        _toggleEndorsements(msg.sender, discordIds);
+    }
+
+    /**
+     * @notice Withdraw for another user (gasless for the vault token holder)
+     * (assuming the owner has given authority for the caller to act on their behalf)
+     *
+     * @dev amount is explicit, to allow use case of partial vault withdrawals
+     */
+    function toggleEndorsementsFor(address account, uint256[] memory discordIds, uint256 deadline, uint8 v, bytes32 r, bytes32 s) public {
+        require(block.timestamp <= deadline, "MasterCandidate: expired deadline");
+
+        bytes32 structHash = keccak256(abi.encode(TOGGLE_ENDORSEMENTS_FOR_TYPEHASH, account, discordIds, deadline, _useNonce(account)));
+        bytes32 digest = _hashTypedDataV4(structHash);
+        address signer = ECDSA.recover(digest, v, r, s);
+
+        require(signer == account, "MasterCandidate: invalid signature");
+
+        _toggleEndorsements(account, discordIds);
+    }
+
+    function tokenURI(uint256 discordId) public view override returns (string memory) {
+        require(_exists(discordId), "ERC721Metadata: URI query for nonexistent token");
+
+        Candidate storage candidate = candidates[discordId];
+
+        // TODO(butlerji): Actually URI encode this somehow?
+        return string(abi.encodePacked(candidate.discordHandle, candidate.twitterHandle, discordId.toString()));
```

Comment:
Undying love and admiration if someone can point me some example code to uri data encode this info (current implementation is just a placeholder)

Reply:
See tokenURI on the LOOT contract - https://etherscan.io/address/0xff9c1b15b16263c61d017ee9f65c50e4ae0113d7#code